spec/requests/custom_field_cache_spec.rb

```diff
@@ -0,0 +1,129 @@
+
+    custom_field.save!
+  end
+
+  def expect_group_by_all_to_include(custom_field)
+    expect(CostQuery::GroupBy.all).to include(group_by_class_name_string(custom_field).constantize)
+  end
+
+  def expect_filter_all_to_include(custom_field)
+    expect(CostQuery::Filter.all).to include(filter_class_name_string(custom_field).constantize)
+  end
+
+  def expect_group_by_all_to_not_exist(custom_field)
+    # can not check for whether the element is included in CostQuery::GroupBy if it does not exist
+    expect { group_by_class_name_string(custom_field).constantize }.to raise_error NameError
+  end
+
+  def expect_filter_all_to_not_exist(custom_field)
+    # can not check for whether the element is included in CostQuery::GroupBy if it does not exist
+    expect { group_by_class_name_string(custom_field).constantize }.to raise_error NameError
+  end
+
+  def visit_cost_reports_index
+    get "projects/#{project.id}/cost_reports"
+  end
+
+  it 'removes the filter/group_by if the custom field is removed' do
+    custom_field2.save!
+
+    visit_cost_reports_index
+
+    expect_group_by_all_to_include(custom_field)
+    expect_group_by_all_to_include(custom_field2)
+
+    expect_filter_all_to_include(custom_field)
+    expect_filter_all_to_include(custom_field2)
+
+    custom_field2.destroy
+
+    visit_cost_reports_index
+
+    expect_group_by_all_to_include(custom_field)
+    expect_group_by_all_to_not_exist(custom_field2)
+
+    expect_filter_all_to_include(custom_field)
+    expect_filter_all_to_not_exist(custom_field2)
+  end
+
+  it 'removes the filter/group_by if the last custom field is removed' do
+    visit_cost_reports_index
+
+    expect_group_by_all_to_include(custom_field)
+    expect_filter_all_to_include(custom_field)
+
+    custom_field.destroy
+
+    visit_cost_reports_index
+
+    expect_group_by_all_to_not_exist(custom_field)
+    expect_filter_all_to_not_exist(custom_field)
+  end
+
+  it 'allows for changing the db table between requests if no caching is done' do
+    old_table_name = WorkPackageCustomField.table_name
+    new_table_name = 'custom_fields_clone'
+    new_id = custom_field.id + 1
+
+    begin
+      mock_cache_classes_setting_with(false)
+
+      visit_cost_reports_index
+
+      expect_group_by_all_to_include(custom_field)
+      expect_filter_all_to_include(custom_field)
+
+      ActiveRecord::Base.connection.execute("CREATE TABLE #{new_table_name} AS SELECT * from custom_fields;")
+      ActiveRecord::Base.connection.execute("UPDATE #{new_table_name} SET id = #{new_id} WHERE id = #{custom_field.id};")
+      CustomField::Translation.where(custom_field_id: custom_field.id).update_all(custom_field_id: new_id)
+
+      WorkPackageCustomField.table_name = new_table_name
+
+      visit_cost_reports_index
+
+      expect_group_by_all_to_not_exist(custom_field)
+      expect_filter_all_to_not_exist(custom_field)
+
+      expect_group_by_all_to_include(new_id)
+      expect_filter_all_to_include(new_id)
```

Comment:
I don't understand how this works.

Usually you pass a `CustomField`, now you pass an `int` and it still does the job?